src/bin/sccache-dist/main.rs

```diff
@@ -99,72 +201,188 @@ fn main() {
+            Ok(0)
+        },
+
+        Command::Scheduler(scheduler_config::Config { client_auth, server_auth }) => {
+            let check_client_auth: Box<Fn(&str) -> bool + Send + Sync> = match client_auth {
+                scheduler_config::ClientAuth::Insecure => Box::new(move |s| s == INSECURE_DIST_CLIENT_TOKEN),
+                scheduler_config::ClientAuth::Token { token } => Box::new(move |s| s == token),
+            };
+
+            let check_server_auth: Box<Fn(&str) -> Option<ServerId> + Send + Sync> = match server_auth {
+                scheduler_config::ServerAuth::Insecure => {
+                    let token = INSECURE_DIST_SERVER_TOKEN;
+                    Box::new(move |server_token| check_server_token(server_token, &token))
+                },
+                scheduler_config::ServerAuth::Token { token } => {
+                    Box::new(move |server_token| check_server_token(server_token, &token))
+                },
+                scheduler_config::ServerAuth::JwtHS256 { secret_key } => {
+                    let secret_key = base64::decode_config(&secret_key, base64::URL_SAFE_NO_PAD).chain_err(|| "Secret key base64 invalid")?;
+                    if secret_key.len() != 256 / 8 {
+                        bail!("Size of secret key incorrect")
+                    }
+                    let validation = jwt::Validation::new(jwt::Algorithm::HS256);
+                    Box::new(move |server_token| check_jwt_server_token(server_token, &secret_key, &validation))
+                }
+            };
+
             let scheduler = Scheduler::new();
-            let http_scheduler = dist::http::Scheduler::new(scheduler);
+            let http_scheduler = dist::http::Scheduler::new(scheduler, check_client_auth, check_server_auth);
             let _: Void = http_scheduler.start();
         },
-        Command::Server { builder, cache_dir, toolchain_cache_size, scheduler_addr } => {
+
+        Command::Server(server_config::Config { builder, cache_dir, public_addr, scheduler_addr, scheduler_auth, toolchain_cache_size }) => {
             let builder: Box<dist::BuilderIncoming<Error=Error>> = match builder {
-                BuilderType::Docker => Box::new(build::DockerBuilder::new()),
-                BuilderType::Overlay { ref bwrap_path, ref build_dir } =>
-                    Box::new(build::OverlayBuilder::new(bwrap_path, build_dir)?)
+                server_config::BuilderType::Docker => Box::new(build::DockerBuilder::new()),
+                server_config::BuilderType::Overlay { bwrap_path, build_dir } =>
+                    Box::new(build::OverlayBuilder::new(bwrap_path, build_dir).chain_err(|| "Overlay builder failed to start")?)
             };
+
+            let server_id = ServerId(public_addr);
+            let scheduler_auth = match scheduler_auth {
+                server_config::SchedulerAuth::Insecure => {
+                    create_server_token(server_id, &INSECURE_DIST_SERVER_TOKEN)
```

Comment:
I think it'd be good to log a warning on server startup for the insecure case.

Reply:
Good point.